### PR TITLE
Feature/api/refactor create hbm

### PR DESCRIPTION
**Pull request by yunheli**

### CODE REVIEW

File: platform-api/src/main/java/com/flow/platform/api/dao/NodeResultDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Override
+    String getKeyName() {
+        return "nodeResultKey";
+    }
+
+
+
+//    @Override
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/dao/NodeResultDaoImpl.java]: delete

File: platform-api/src/main/java/com/flow/platform/api/dao/hbm/job.hbm.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        <param name="useNamed">true</param>
+      </type>
+    </property>
+
+  </class>
+
+
+  <class name="JobYmlStorage" table="job_yml_storage">
+    <id name="jobId" type="big_integer">
+      <column name="job_id" sql-type="decimal(25, 2)"></column>
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/dao/hbm/job.hbm.xml]: 2 -> 0

File: platform-api/src/main/java/com/flow/platform/api/service/JobServiceImpl.java [hunk 3]
```diff
@@ -254,13 +274,15 @@ private void sessionCallback(Job job, CmdBase cmdBase) {
             job.setSessionId(cmdBase.getSessionId());
             update(job);
             // run step
-            JobFlow jobFlow = (JobFlow) jobNodeService.find(job.getNodePath());
-            if (jobFlow == null) {
-                throw new NotFoundException(String.format("Not Found Job Flow - %s", jobFlow.getName()));
+            NodeResult nodeResult = nodeResultService.find(job.getNodePath(), job.getId());
+            Node flow = jobYmlStorageService.get(job.getId(), nodeResult.getNodeResultKey().getPath());
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/service/JobServiceImpl.java]: jobYmlStorageService.get(nodeResult.getNodeResultKey())

File: platform-api/src/main/java/com/flow/platform/api/service/JobServiceImpl.java [hunk 4]
```diff
@@ -269,40 +291,46 @@ private void sessionCallback(Job job, CmdBase cmdBase) {
...
 
     /**
      * update job flow status
      */
-    private void updateJobStatus(JobNode jobNode) {
-        Job job = jobNode.getJob();
-        if (job == null) {
+    private void updateJobStatus(NodeResult nodeResult) {
+        Node node = jobYmlStorageService
+            .get(nodeResult.getNodeResultKey().getJobId(), nodeResult.getNodeResultKey().getPath());
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/service/JobServiceImpl.java]: change

File: platform-api/src/main/java/com/flow/platform/api/service/JobYmlStorageService.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import java.math.BigInteger;
+
+/**
+ * @author lhl
+ */
+public interface JobYmlStorageService {
+
+    void save(BigInteger jobId, String yml);
+
+    Node get(BigInteger jobId);
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/service/JobYmlStorageService.java]: delete

File: platform-api/src/test/java/com/flow/platform/api/test/controller/WebhookControllerTest.java [hunk 6]
```diff
@@ -57,40 +68,35 @@
...
-                .withBody(cmdRes.toJson())));
-
-        stubFor(com.github.tomakehurst.wiremock.client.WireMock.post(urlEqualTo("/cmd/send"))
-            .willReturn(aResponse()
-                .withBody(cmdRes.toJson())));
-    }
+    @Autowired
+    private YmlStorageDao ymlStorageDao;
+
+//    private void stubDemo() {
```
Comment (yunheli) [inline platform-api/src/test/java/com/flow/platform/api/test/controller/WebhookControllerTest.java]: delete

File: platform-api/src/main/java/com/flow/platform/api/controller/JobController.java [hunk 7]
```diff
@@ -47,20 +50,29 @@
...
 
-    @PostMapping(path = "/jobs")
+    @Autowired
+    private YmlStorageService ymlStorageService;
+
+    @PostMapping
     public Job create(@RequestBody String body) {
         Node node = NodeUtil.buildFromYml(body);
         nodeService.create(node);
+        ymlStorageService.save(node.getPath(), body);
```
Comment (stargwq) [inline platform-api/src/main/java/com/flow/platform/api/controller/JobController.java]: put those into job service

File: platform-api/src/main/java/com/flow/platform/api/controller/JobController.java [hunk 8]
```diff
@@ -47,20 +50,29 @@
...
-    @GetMapping(path = "/jobs/{id}")
-    public Job show(@PathVariable String id) {
+    @GetMapping(path = "/{id}")
+    public Job show(@PathVariable BigInteger id) {
         return jobService.find(id);
     }
+
+    @PostMapping(path = "/status/latest")
+    public List<Job> listLatestJobStatus(@RequestBody List<String> flowNames){
+        return jobService.listLatestJobs(flowNames);
```
Comment (stargwq) [inline platform-api/src/main/java/com/flow/platform/api/controller/JobController.java]: same as list job

File: platform-api/src/main/java/com/flow/platform/api/controller/JobController.java [hunk 9]
```diff
@@ -47,20 +50,29 @@
...
 
-    @GetMapping(path = "/jobs/{id}")
-    public Job show(@PathVariable String id) {
+    @GetMapping(path = "/{id}")
+    public Job show(@PathVariable BigInteger id) {
         return jobService.find(id);
     }
+
+    @PostMapping(path = "/status/latest")
+    public List<Job> listLatestJobStatus(@RequestBody List<String> flowNames){
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/controller/JobController.java]: delete

File: platform-api/src/main/java/com/flow/platform/api/dao/FlowDao.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * @author yh@firim
+ */
+public interface FlowDao extends BaseDao<String, Flow> {
+
+    /**
+     * list flows
+     */
+    List<Flow> list();
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/dao/FlowDao.java]: add basedao

File: platform-api/src/main/java/com/flow/platform/api/dao/FlowDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    String getKeyName() {
+        return "path";
+    }
+
+    @Override
+    public List<Flow> list() {
+        return execute((Session session) -> {
```
Comment (yunheli) [inline platform-api/src/main/java/com/flow/platform/api/dao/FlowDaoImpl.java]: change to common

File: platform-api/src/main/java/com/flow/platform/api/domain/NodeResult.java [hunk 12]
```diff
@@ -110,11 +164,11 @@ public void setStartTime(ZonedDateTime startTime) {
...
-        return cmdId;
+    public ZonedDateTime getFinishTime() {
+        return finishTime;
     }
 
-    public void setCmdId(String cmdId) {
-        this.cmdId = cmdId;
+    public void setFinishTime(ZonedDateTime finishTime) {
+        this.finishTime = finishTime;
     }
```
Comment (stargwq) [inline platform-api/src/main/java/com/flow/platform/api/domain/NodeResult.java]: equal and hashcode

File: platform-api/src/main/java/com/flow/platform/api/service/AgentServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        StringBuilder stringBuilder = new StringBuilder(agentUrl);
+        stringBuilder.append("?zone=").append(zone);
+        String res = HttpUtil.get(stringBuilder.toString());
+        if (res == null) {
+            throw new RuntimeException("Get Agent List error");
+        }
+
+        Agent[] agents = Jsonable.GSON_CONFIG.fromJson(res, Agent[].class);
+        List<AgentWithFlow> agentWithFlows = new ArrayList<>();
+        List<Job> jobs = jobDao.list(NodeStatus.RUNNING);
```
Comment (stargwq) [inline platform-api/src/main/java/com/flow/platform/api/service/AgentServiceImpl.java]: where in query

File: platform-api/src/main/java/com/flow/platform/api/service/NodeServiceImpl.java [hunk 14]
```diff
@@ -50,11 +63,25 @@ public Node create(Node node) {
     @Override
     public Node save(Node node) {
         mocNodeList.put(node.getPath(), node);
```
Comment (stargwq) [inline platform-api/src/main/java/com/flow/platform/api/service/NodeServiceImpl.java]: guava cache

File: platform-api/src/main/java/com/flow/platform/api/service/NodeServiceImpl.java [hunk 15]
```diff
@@ -33,15 +38,23 @@
...
-             String env = System.getProperty("flow.api.env");
+            String env = System.getProperty("flow.api.env");
             if (item instanceof Step && env != "test") {
                 if (NodeUtil.canRun(item) && (item.getScript() == null || item.getScript().isEmpty())) {
-                    throw new IllegalParameterException(String.format("Missing Param Script, NodeName: %s", item.getName()));
+                    throw new IllegalParameterException(
+                        String.format("Missing Param Script, NodeName: %s", item.getName()));
                 }
             }
+
```
Comment (stargwq) [inline platform-api/src/main/java/com/flow/platform/api/service/NodeServiceImpl.java]: check flow

**Pull request merged**